sklearn/utils/tests/test_validation.py

```diff
@@ -974,25 +976,54 @@ def test_check_is_fitted_with_attributes(wrap):
 
 
 def test_check_consistent_length():
+    """Test that `check_consistent_length` raises on inconsistent lengths and wrong
+    input types trigger TypeErrors."""
     check_consistent_length([1], [2], [3], [4], [5])
     check_consistent_length([[1, 2], [[1, 2]]], [1, 2], ["a", "b"])
     check_consistent_length([1], (2,), np.array([3]), sp.csr_matrix((1, 2)))
     with pytest.raises(ValueError, match="inconsistent numbers of samples"):
         check_consistent_length([1, 2], [1])
+
     with pytest.raises(TypeError, match=r"got <\w+ 'int'>"):
         check_consistent_length([1, 2], 1)
     with pytest.raises(TypeError, match=r"got <\w+ 'object'>"):
         check_consistent_length([1, 2], object())
-
     with pytest.raises(TypeError):
         check_consistent_length([1, 2], np.array(1))
-
     # Despite ensembles having __len__ they must raise TypeError
     with pytest.raises(TypeError, match="Expected sequence or array-like"):
         check_consistent_length([1, 2], RandomForestRegressor())
     # XXX: We should have a test with a string, but what is correct behaviour?
 
 
+@pytest.mark.parametrize(
+    "array_namespace, device, _", yield_namespace_device_dtype_combinations()
+)
+def test_check_consistent_length_array_API(array_namespace, device, _):
+    """Test that check_consistent_length works with different array types."""
+    xp = _array_api_for_tests(array_namespace, device)
+
+    check_consistent_length(
+        xp.asarray([1], device=device),
+        xp.asarray([2], device=device),
+    )
+    if xp.__name__ == "numpy":
+        check_consistent_length(
+            xp.asarray([[1, 2], [1, 2]], device=device),
+            xp.asarray([1, 2], device=device),
+            xp.asarray(["a", "b"], device=device),
```

Comment:
Since the array API spec does not include string values I would not include this case in this function to avoid implying otherwise.

Here is [item 4 of the "out of scope" section of the array API spec](https://data-apis.org/array-api/latest/purpose_and_scope.html):

> 4. Standardization of these dtypes is out of scope: bfloat16, extended precision floating point, datetime, string, object and void dtypes.

> Rationale: these dtypes aren’t uniformly supported, and their inclusion at this point in time could put a significant implementation burden on libraries. It is expected that some of these dtypes - in particular bfloat16 - will be included in a future version of the standard.

Testing for string value inputs is already done in existing tests, so there is no need to include those cases in `test_check_consistent_length_array_api`.

As a result there is also no need to keep the `if xp.__name__ == "numpy"` branch under `test_check_consistent_length_array_api`.

On the other hand, I think we should check that `check_consistent_length` can check that any array API input has a consistent length with both array API inputs and any non array API input such as a Python list of strings or a numpy array of strings, scipy sparse matrices, (or even pandas or polars dataframes):


```python
with config_context(array_api_dispatch=True):
    check_consistent_length(
        xp.asarray([1, 2, 3], device=device),
        xp.asarray([[1, 1], [2, 2], [3, 3]], device=device),
        [1, 2, 3],
        ["a", "b", "c"],
        np.asarray(["a", "b", "c", dtype=object),
        sp.csr_array([[0, 1], [1, 0], [0, 0]]),
    )
```